services/github/src/handlers/status.js

```diff
@@ -1,30 +1,221 @@
-const { CONCLUSIONS, CHECKLOGS_TEXT, CHECKRUN_TEXT, LIVE_LOG_ARTIFACT_NAME, CUSTOM_CHECKRUN_TEXT_ARTIFACT_NAME, CUSTOM_CHECKRUN_ANNOTATIONS_ARTIFACT_NAME, CHECK_RUN_STATES, TASK_STATE_TO_CHECK_RUN_STATE } = require('../constants');
-const { tailLog } = require('../utils');
+const { CONCLUSIONS, CHECKLOGS_TEXT, CHECKRUN_TEXT, LIVE_LOG_ARTIFACT_NAME,
+  CUSTOM_CHECKRUN_TEXT_ARTIFACT_NAME, CUSTOM_CHECKRUN_ANNOTATIONS_ARTIFACT_NAME,
+  CHECK_RUN_STATES, TASK_STATE_TO_CHECK_RUN_STATE,
+} = require('../constants');
+const { tailLog, markdownLog, markdownAnchor } = require('../utils');
 const { requestArtifact } = require('./requestArtifact');
 const { taskUI, makeDebug, taskLogUI } = require('./utils');
 
+let counter = 0; // unique event counter for logs, to group related log messages
+const taskEventsMap = new Map();
 /**
- * Post updates to GitHub, when the status of a task changes. Uses Checks API
+ * Tracking events order to prevent older events from overwriting newer updates
+ * Rabbitmq messages will be arriving strictly in the order they were created,
+ * but the consumer itself doesn't wait for the previous handler to finish,
+ * if there is a new message coming in.
+ * As handlers are async, and are doing external API calls and db calls,
+ * they can spend different amount of time before sending final create/update checkrun.
+ * This can lead to situations where newer event is being overwritten by an older one,
+ * because it took him longer to reach update calls.
+ *
+ * We simply store the time when handler started. Code can use `isLatest()`
+ * to check if it is still the latest handler (by time) and abort if it is not.
+ * `taskFinished()` should be used to clean up own time reference.
+ *
+ * @param {string} taskId
+ * @returns {object} { isLatest, taskFinished }
+ */
+const taskStarted = (taskId) => {
+  const now = new Date().getTime();
+  taskEventsMap.set(taskId, now);
+
+  const isLatest = () => taskEventsMap.get(taskId) <= now;
+  const taskFinished = () => {
+    if (taskEventsMap.get(taskId) === now) {
+      // only remove entry if it was the one creating it
+      taskEventsMap.delete(taskId);
+    }
+  };
+  return { isLatest, taskFinished };
+};
+
+class GithubCheckOutput {
+  constructor({
+    title = '',
+    summary = '',
+    text = '',
+    annotations = [],
+  }) {
+    this.title = title;
+    this.summary = summary;
+    this.text = text;
+    this.annotations = annotations;
+  }
+
+  addText(text, appendText = '\n') {
+    this.text += `${text}${appendText}`;
+  }
+
+  /**
+   * Github has a limit of 64Kb for the whole payload
+   */
+  getRemainingMaxSize() {
```

Comment:
this will ensure that if someone includes custom annotations, we will consider its size too